### PR TITLE
Make C++ TM and python TM produce identical results

### PR DESCRIPTION
**Pull request by chetan51**
Fixes https://github.com/numenta/nupic/issues/2964.

Depends on https://github.com/numenta/nupic.core/pull/838.

### CODE REVIEW

File: src/nupic/research/temporal_memory.py [hunk 1]
```diff
@@ -285,7 +285,8 @@ def burstColumns(self,
 
     unpredictedActiveColumns = activeColumns - predictedActiveColumns
 
-    for column in unpredictedActiveColumns:
+    # Sort unpredictedActiveColumns before iterating for compatibility with C++
+    for column in sorted(list(unpredictedActiveColumns)):
```
Comment (scottpurdy) [inline src/nupic/research/temporal_memory.py]: `sorted` will work on a set so no need to cast to a list first

File: src/nupic/research/temporal_memory.py [hunk 2]
```diff
@@ -340,7 +341,12 @@ def learnOnSegments(self,
     @param predictedInactiveCells       (set)         Indices of predicted inactive cells
     @param prevMatchingSegments         (set)         Indices of segments with
     """
-    for segment in prevActiveSegments | learningSegments:
+    # Sort segments before iterating for compatibility with C++
+    segments = prevActiveSegments | learningSegments
+    segments = sorted(list(segments),
+                      key=lambda segment: connections.cellForSegment(segment))
```
Comment (scottpurdy) [inline src/nupic/research/temporal_memory.py]: Can multiple segments belong to the same cell? If so, then we may need a secondary key here to make sure segments for the same cell have a strict ordering.

**Comment by scottpurdy** [comment]
Done reviewing

**Comment by rhyolight** [comment]
Sorry @chetan51 the CI failure is my fault. I'm working on it.

**Comment by rhyolight** [comment]
Actually, I can [blame Travis-CI for this](https://www.traviscistatus.com/incidents/ccjlfnfbhfdh), I think.

**Comment by chetan51** [comment]
@scottpurdy Added a secondary sort in this PR and in https://github.com/numenta/nupic.core/pull/838/files#diff-aeec0528733ed99b3d31a8b4973a8726R350.

File: src/nupic/research/temporal_memory.py [hunk 3]
```diff
@@ -340,7 +341,16 @@ def learnOnSegments(self,
...
     @param prevMatchingSegments         (set)         Indices of segments with
     """
-    for segment in prevActiveSegments | learningSegments:
+    segments = prevActiveSegments | learningSegments
+
+    # Sort segments before iterating for compatibility with C++
+    # Sort on secondary key (segment idx)
+    segments = sorted(segments)
+    # Sort on primary key (cell idx)
+    segments = sorted(segments,
```
Comment (scottpurdy) [inline src/nupic/research/temporal_memory.py]: This is assuming a stable sort to preserve the secondary sort key from the previous sort? You should instead combine the sorts by having the lambda function return a tuple `(primaryKey, secondaryKey)`

**Comment by scottpurdy** [comment]
Done reviewing

File: src/nupic/research/temporal_memory.py (hunk 3 above)
Comment (chetan51) [inline src/nupic/research/temporal_memory.py]: I followed [this](https://docs.python.org/2/howto/sorting.html#sort-stability-and-complex-sorts), but I like the idea of using a tuple instead.

**Comment by chetan51** [comment]
@scottpurdy Updated.

**Comment by chetan51** [comment]
@rhyolight AppVeyor build keeps getting cancelled.

**Comment by rhyolight** [comment]
This is probably my fault. Please wait until after I've done this release.

File: src/nupic/research/temporal_memory.py [hunk 4]
```diff
@@ -340,7 +341,14 @@ def learnOnSegments(self,
     @param predictedInactiveCells       (set)         Indices of predicted inactive cells
     @param prevMatchingSegments         (set)         Indices of segments with
     """
-    for segment in prevActiveSegments | learningSegments:
+    segments = prevActiveSegments | learningSegments
+
+    # Sort segments before iterating for compatibility with C++
+    # Sort with primary key = cell idx, secondary key = segment idx
+    segments = sorted(segments,
```
Comment (scottpurdy) [inline src/nupic/research/temporal_memory.py]: nit: drop `segments` argument to next line with same indentation as `key` argument

**Comment by scottpurdy** [comment]
Done reviewing, just one style nit pick, :+1: , merge when ready

**Pull request merged**